src/jsMain/kotlin/baaahs/sim/ui/PinkyPanel.kt

```diff
@@ -0,0 +1,132 @@
+package baaahs.sim.ui
+
+import baaahs.Pinky
+import baaahs.ui.BComponent
+import baaahs.ui.Observable
+import baaahs.ui.Observer
+import baaahs.ui.SimulatorStyles
+import baaahs.ui.SimulatorStyles.beatsDiv
+import baaahs.ui.SimulatorStyles.showsDiv
+import baaahs.util.percent
+import kotlinx.html.id
+import kotlinx.html.js.onChangeFunction
+import org.w3c.dom.HTMLSelectElement
+import react.*
+import react.dom.*
+import styled.css
+import styled.styledDiv
+import styled.styledSelect
+import styled.styledSpan
+import kotlin.math.roundToInt
+
+class PinkyPanel(props: Props) : BComponent<PinkyPanel.Props, PinkyPanel.State>(props), Observer {
+    override fun observing(props: Props, state: State): List<Observable?> {
+        return listOf(props.pinky)
+    }
+
+    inline fun <reified T : RComponent<P, *>, reified P: RProps> RBuilder.render(
+        props: P,
+        noinline block: RBuilder.() -> Unit = {}
+    ) {
+        child(createElement(T::class.js, props, block))
+    }
+
+    override fun RBuilder.render() {
+        val pinky = props.pinky
+        val shows = pinky.shows
+
+        styledDiv {
+            css { +SimulatorStyles.section }
+            b { +"Pinky:" }
+            div {
+                attrs.id = "pinkyView" // TODO: kill
+                +"Current Show: "
+
+                styledSelect {
+                    css { +showsDiv }
+                    attrs.onChangeFunction = { event ->
+                        println("event.target = ${event.target}")
+                        val select = event.target as HTMLSelectElement
+                        val nextShow = shows[select.value.toInt()]
+                        pinky.selectedShow = nextShow
+                    }
+
+                    val selectedShow = pinky.selectedShow
+                    shows.forEachIndexed { index, show ->
+                        option {
+                            +show.name
+                            attrs.key = index.toString()
+                            attrs.value = index.toString()
+                            if (selectedShow.name == show.name) {
+                                attrs.selected = true
+                            }
+                        }
+                    }
+                }
+            }
+
+            br {  }
+            +"Brains online: "
+
+            span {
+                +"${pinky.brains.size}"
+            }
+
+            styledDiv {
+                css { +beatsDiv }
+
+                val beatData = pinky.beatData
+                val beat = beatData.beatWithinMeasure(pinky.clock).toInt()
+                val bpm = beatData.bpm
+                val beatConfidence = beatData.confidence
+
+                b { +"Beats: " }
+                span { // beatConfidenceElement
+                    +"[confidence: ${beatConfidence.percent()}]"
+                }
+                br { }
+                styledDiv { +"1"; if (beat == 0) css { +"selected" } }
```

Comment:
This syntax is really interesting and I don't know whether to be appalled or maybe you're onto something here?

Are there any links to documentation/discussion threads about what looks like sass but kinda like jsx but totally kotlin? 

I still will ask: what is wrong with using jsx?

Reply:
Heheh yeah I'm not used to it yet either.

`styledDiv` is gross; it's the same as `div` with access to CSS config.

I'm using Kotlin for the React rendering here cuz I really like having edit-time type checking, autocomplete, and refactoring. It compiles into CSS just like JSX, so, same same.

Reply:
(and JSX still has the bewildering panoply of JS idioms and historical brokenness (`function(){}.bind(this)` et al.), which Kotlin fixes.)

Reply:
Fine with me if you wanna use JSX here though, this is just easier/more familiar to me.

Reply:
But to answer your question, it's equivalent to:

```html
<div class={beat == 0 ? "selected" : ""}>1</div>
```

Some Kotlin DSLs use the `+` operator as a sort of "contextually add this thing" command. So inside a tag it's _add string_; inside a CSS block it's _add class_.